File: src/src_dotnet/JAStudio.Core/Note/Collection/NoteCache.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Compze.Utilities.SystemCE.LinqCE;
using Compze.Utilities.SystemCE.ThreadingCE.ResourceAccess;
using JAStudio.PythonInterop;

namespace JAStudio.Core.Note.Collection;

internal class CachedNote
{
   public NoteId Id { get; }
   public string Question { get; }

   public CachedNote(JPNote note)
   {
      Id = note.GetId();
      Question = note.GetQuestion();
   }
}

internal abstract class NoteCacheBase<TNote> : IExternalNoteUpdateHandler where TNote : JPNote
{
   readonly Func<NoteServices, NoteData, TNote> _noteConstructor;
   readonly Type _noteType;
   protected readonly Dictionary<NoteId, TNote> _byId = new();
   readonly List<Action<TNote>> _updateListeners = new();
   readonly NoteServices _noteServices;
   protected readonly IMonitorCE _monitor = IMonitorCE.WithDefaultTimeout();

   protected NoteCacheBase(Type cachedNoteType, Func<NoteServices, NoteData, TNote> noteConstructor, NoteServices noteServices)
   {
      _noteConstructor = noteConstructor;
      _noteType = cachedNoteType;
      _noteServices = noteServices;
   }

   public void OnNoteUpdated(dynamic listener)
   {
      Action<TNote> callback = PythonDotNetShim.Action.ToDotNet<TNote>(listener);
      _updateListeners.Add(callback);
   }

   public void OnNoteUpdated(Action<TNote> listener)
   {
      _updateListeners.Add(listener);
   }

   public TNote? WithIdOrNone(NoteId noteId) => _monitor.Read(() => WithIdOrNoneCore(noteId));
   protected TNote? WithIdOrNoneCore(NoteId noteId) => _byId.TryGetValue(noteId, out var note) ? note : null;

   /// <summary>Look up a note by its external long ID (uses the shared ExternalNoteIdMap).</summary>
   public TNote? WithExternalIdOrNone(long externalNoteId) =>
      _noteServices.ExternalNoteIdMap.FromExternalId(externalNoteId) is {} noteId ? WithIdOrNone(noteId) : null;

   TNote? WithExternalIdOrNoneCore(long externalNoteId) =>
      _noteServices.ExternalNoteIdMap.FromExternalId(externalNoteId) is {} noteId ? WithIdOrNoneCore(noteId) : null;

   /// <summary>Converts an external long ID to the corresponding domain NoteId.</summary>
   public NoteId? ExternalIdToNoteId(long externalNoteId) =>
      _noteServices.ExternalNoteIdMap.FromExternalId(externalNoteId);

   /// <summary>Returns the external long note ID for the given domain NoteId.</summary>
   public long GetExternalNoteId(NoteId noteId) =>
      _noteServices.ExternalNoteIdMap.ToExternalId(noteId) ?? 0;

   public void ExternalNoteAdded(long externalNoteId, NoteData data)
   {
      data.Id = CreateTypedId(Guid.NewGuid());
      _noteServices.ExternalNoteIdMap.Register(externalNoteId, data.Id);
      var note = _noteConstructor(_noteServices, data);
      note.UpdateGeneratedData();
      AddToCache(note);
   }

   // ReSharper disable once FieldCanBeMadeReadOnly.Local
   bool _ankiStillOwnsNoFields = true;

   public void ExternalNoteWillFlush(long externalNoteId, NoteData data)
   {
      // TODO: Currently disabled: Anki owns no note fields (all authoritative data is in the file system repository). Consider whether to remove this and related code entirely
      if(_ankiStillOwnsNoFields) //Keeping the code below around and not detected as unreachable through this little hack
         return;

      var noteId = _noteServices.ExternalNoteIdMap.FromExternalId(externalNoteId);
      if(noteId == null) return;
      var existing = WithIdOrNone(noteId);
      if(existing == null) return;
      if(existing.IsFlushing) return; // Our code initiated this flush, nothing to do

      data.Id = noteId;
      var note = CreateNoteByMergingAnkiData(_noteServices, existing, data);
      note.CopyStudyingStatusFrom(existing);
      note.UpdateGeneratedData();
      _monitor.Read(() => RefreshInCacheCore(note));
      NotifyUpdateListeners(note);
   }

   public void ExternalNoteRemoved(long externalNoteId)
   {
      var noteId = _noteServices.ExternalNoteIdMap.FromExternalId(externalNoteId);
      if(noteId == null) return;
      _monitor.Read(() =>
      {
         var existing = WithIdOrNoneCore(noteId);
         if(existing != null)
         {
            RemoveFromCacheCore(existing);
         }
      });

      _noteServices.ExternalNoteIdMap.Unregister(externalNoteId);
   }

   /// <summary>Creates the correctly typed NoteId for this cache (VocabId, KanjiId, etc.)</summary>
   protected abstract NoteId CreateTypedId(Guid value);

   /// <summary>Creates a new note by merging Anki-owned fields into the existing note's data, preserving all fields Anki does not store.</summary>
   protected abstract TNote CreateNoteByMergingAnkiData(NoteServices services, TNote existing, NoteData ankiData);

   public void JpNoteUpdated(TNote note)
   {
      _monitor.Read(() =>
      {
         var existing = WithIdOrNoneCore(note.GetId());
         if(existing == null)
            throw new InvalidOperationException($"JpNoteUpdated called for {typeof(TNote).Name} with id {note.GetId()} but it is not in the cache. Only persisted (cached) notes should flush.");

         RefreshInCacheCore(note);
      });

      NotifyUpdateListeners(note);
   }

   void NotifyUpdateListeners(TNote note)
   {
      foreach(var listener in _updateListeners)
      {
         listener(note);
      }
   }

   void RefreshInCacheCore(TNote note)
   {
      RemoveFromCacheCore(note);
      AddToCacheCore(note);
   }

   /// <summary>Removes all notes and ID mappings from the cache. Subclasses must override ClearInheritorIndexes to clear their custom indexes.</summary>
   public void Clear() => _monitor.Read(() =>
   {
      _byId.Clear();
      ClearInheritorIndexes();
   });

   protected abstract void ClearInheritorIndexes();

   public void AddToCache(TNote note) => _monitor.Read(() => AddToCacheCore(note));
   public void AddAllToCache(IEnumerable<TNote> notes) => _monitor.Read(() => notes.ForEach(AddToCacheCore));
   public void RemoveFromCache(TNote note) => _monitor.Read(() => RemoveFromCacheCore(note));

   protected abstract void AddToCacheCore(TNote note);
   protected abstract void RemoveFromCacheCore(TNote note);

   public void SetStudyingStatuses(Dictionary<long, List<CardStudyingStatus>> statusesByExternalId)
   {
      _monitor.Read(() =>
      {
         foreach(var (externalId, statuses) in statusesByExternalId)
         {
            var note = WithExternalIdOrNoneCore(externalId);
            if(note != null)
            {
               foreach(var status in statuses)
                  note.SetStudyingStatus(status);
            }
         }
      });
   }
}

internal abstract class NoteCache<TNote, TSnapshot> : NoteCacheBase<TNote>
   where TNote : JPNote
   where TSnapshot : CachedNote
{
   readonly Dictionary<string, HashSet<TNote>> _byQuestion = new();
   readonly Dictionary<NoteId, TSnapshot> _snapshotById = new();

   protected NoteCache(Type cachedNoteType, Func<NoteServices, NoteData, TNote> noteConstructor, NoteServices noteServices)
      : base(cachedNoteType, noteConstructor, noteServices) {}

   public List<TNote> All() => _monitor.Read(() => _byId.Values.ToList());

   public List<TNote> WithQuestion(string question) => _monitor.Read(() => _byQuestion.TryGetValue(question, out var notes) ? notes.ToList() : new List<TNote>());

   protected abstract TSnapshot CreateSnapshot(TNote note);
   protected abstract void InheritorRemoveFromCache(TNote note, TSnapshot snapshot);
   protected abstract void InheritorAddToCache(TNote note, TSnapshot snapshot);

   protected override void ClearInheritorIndexes()
   {
      _byQuestion.Clear();
      _snapshotById.Clear();
      ClearDerivedIndexes();
   }

   /// <summary>Override in leaf caches to clear type-specific secondary indexes.</summary>
   protected abstract void ClearDerivedIndexes();

   protected override void RemoveFromCacheCore(TNote note)
   {
      var id = note.GetId();

      if(!_snapshotById.TryGetValue(id, out var cached))
         throw new InvalidOperationException($"Cannot remove {typeof(TNote).Name} with id {id} from cache: not found in snapshot index");

      _snapshotById.Remove(id);
      _byId.Remove(id);

      if(_byQuestion.TryGetValue(cached.Question, out var questionList))
      {
         questionList.Remove(note);
      }

      InheritorRemoveFromCache(note, cached);
   }

   protected override void AddToCacheCore(TNote note)
   {
      var id = note.GetId();

      if(_snapshotById.ContainsKey(id))
         throw new InvalidOperationException($"Cannot add {typeof(TNote).Name} with id {id} to cache: already present. Use RefreshInCache for updates.");

      var snapshot = CreateSnapshot(note);
      _snapshotById[id] = snapshot;
      _byId[id] = note;

      if(!_byQuestion.ContainsKey(snapshot.Question))
      {
         _byQuestion[snapshot.Question] = new HashSet<TNote>();
      }

      _byQuestion[snapshot.Question].Add(note);

      InheritorAddToCache(note, snapshot);

      note.MarkAsPersisted();
   }
}

File: src/jaslib_src/jaslib/note/collection/note_cache.py
from __future__ import annotations

from collections import defaultdict
from typing import TYPE_CHECKING, override

from autoslot import Slots
from jaspythonutils.sysutils.abstract_method_called_error import AbstractMethodCalledError
from typed_linq_collections.collections.q_list import QList

from jaslib.note.jpnote import JPNote, JPNoteId
from jaslib.task_runners.task_progress_runner import TaskRunner

if TYPE_CHECKING:
    from collections.abc import Callable

    from jaslib.note.collection.card_studying_status import CardStudyingStatus
    from jaslib.note.jpnote_data import JPNoteData

class CachedNote(Slots):
    def __init__(self, note: JPNote) -> None:
        self.id: JPNoteId = note.get_id()
        self.question: str = note.get_question()

class NoteCacheBase[TNote: JPNote](Slots):
    def __init__(self, cached_note_type: type[TNote], note_constructor: Callable[[JPNoteData], TNote]) -> None:
        self._note_constructor: Callable[[JPNoteData], TNote] = note_constructor
        self._note_type: type[TNote] = cached_note_type
        self._by_id: dict[JPNoteId, TNote] = {}

        self._update_listeners: list[Callable[[TNote], None]] = []

    def on_note_updated(self, listener: Callable[[TNote], None]) -> None:
        self._update_listeners.append(listener)

    def with_id_or_none(self, note_id: JPNoteId) -> TNote | None:
        return self._by_id.get(note_id, None)

    def anki_note_updated(self, note: TNote) -> None:
        self._refresh_in_cache(note)

    def jp_note_updated(self, note: TNote) -> None:
        self._refresh_in_cache(note)
        self._notify_update_listeners(note)

    def _notify_update_listeners(self, note: TNote) -> None:
        for listener in self._update_listeners: listener(note)

    def _refresh_in_cache(self, note: TNote) -> None:
        self.remove_from_cache(note)
        self.add_to_cache(note)

    def init_from_list(self, all_notes: list[JPNoteData]) -> None:
        if len(all_notes) > 0:
            with TaskRunner.current(f"Pushing {self._note_type.__name__} notes into cache") as task_runner:
                task_runner.process_with_progress(all_notes, self._add_to_cache_from_data, f"Pushing {self._note_type.__name__} notes into cache")

    def _add_to_cache_from_data(self, note_data: JPNoteData) -> None:
        self.add_to_cache(self._note_constructor(note_data))

    def remove_from_cache(self, note: TNote) -> None: raise AbstractMethodCalledError()  # pyright: ignore[reportUnusedParameter]
    def add_to_cache(self, note: TNote) -> None: raise AbstractMethodCalledError()  # pyright: ignore[reportUnusedParameter]

    def set_studying_statuses(self, card_statuses: list[CardStudyingStatus]) -> None:
        for status in card_statuses:
            note = self.with_id_or_none(status.note_id)
            if note is not None:
                note.set_studying_status(status)

class NoteCache[TNote: JPNote, TSnapshot: CachedNote](NoteCacheBase[TNote], Slots):
    def __init__(self, cached_note_type: type[TNote], note_constructor: Callable[[JPNoteData], TNote]) -> None:
        super().__init__(cached_note_type, note_constructor)
        # Since notes with a given Id are guaranteed to only exist once in the cache, we can use lists within the dictionary to cut memory usage a ton compared to using sets
        self._by_question: defaultdict[str, QList[TNote]] = defaultdict[str, QList[TNote]](QList[TNote])
        self._snapshot_by_id: dict[JPNoteId, TSnapshot] = {}

    def all(self) -> QList[TNote]:
        return QList(self._by_id.values())

    def with_question(self, question: str) -> QList[TNote]:
        if question in self._by_question: return self._by_question[question]
        return QList[TNote]()

    def _create_snapshot(self, note: TNote) -> TSnapshot: raise AbstractMethodCalledError()  # pyright: ignore[reportUnusedParameter]
    def _inheritor_remove_from_cache(self, note: TNote, snapshot: TSnapshot) -> None: raise AbstractMethodCalledError()  # pyright: ignore[reportUnusedParameter]
    def _inheritor_add_to_cache(self, note: TNote, snapshot: TSnapshot) -> None: raise AbstractMethodCalledError()  # pyright: ignore[reportUnusedParameter]

    @override
    def remove_from_cache(self, note: TNote) -> None:
        assert note.get_id()
        cached = self._snapshot_by_id.pop(note.get_id())
        self._by_id.pop(note.get_id())
        self._by_question[cached.question].remove(note)
        self._inheritor_remove_from_cache(note, cached)

    @override
    def add_to_cache(self, note: TNote) -> None:
        if note.get_id() in self._by_id: return
        if note.get_id() == 0:
            raise Exception("Cannot add note with id 0 to cache")

        self._by_id[note.get_id()] = note
        snapshot = self._create_snapshot(note)
        self._snapshot_by_id[note.get_id()] = snapshot
        self._by_question[note.get_question()].append(note)
        self._inheritor_add_to_cache(note, snapshot)
